### ISSUE TITLE
Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap.

### ISSUE THREAD

**Issue by liuzqt** [quarkusio/quarkus#18210]
## Describe the bug

This error happened when I was trying to build native executables
```
Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Trace: 
        at parsing org.testcontainers.utility.Base58.randomString(Base58.java:19)
Call path from entry point to org.testcontainers.utility.Base58.randomString(int): 
        at org.testcontainers.utility.Base58.randomString(Base58.java:16)
        at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:139)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:90)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:82)
        at ai.tusimple.map.feature.AppLifecycleBean.initLocalS3(AppLifecycleBean.java:69)
        at ai.tusimple.map.feature.AppLifecycleBean.onStart(AppLifecycleBean.java:134)
        at ai.tusimple.map.feature.AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.notify(AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.zig:147)
        at io.quarkus.arc.impl.EventImpl$DeferredEventNotification.run(EventImpl.java:427)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
        at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)
```

It seems that it has something to do with my usage with testcontainer and localstack. I used testcontainer to provision a localstack docker image in order to mock S3 environment in dev and test profile, during the `void onStart(@Observes StartupEvent ev)` quarkus startup event. If that's the problem, is there any way to tell quarkus to ignore this part when building production jar and executables? 

A complete error log is below:
```
> Task :feature:quarkusBuild
Caching disabled for task ':feature:quarkusBuild' because:
  Build cache is disabled
Task ':feature:quarkusBuild' is not up-to-date because:
  Task has failed previously.
building quarkus jar
JBoss Threads version 3.2.0.Final

[feature-1.0.0-SNAPSHOT-runner:8250]    classlist:   2,311.89 ms,  1.70 GB
[feature-1.0.0-SNAPSHOT-runner:8250]        (cap):   1,306.53 ms,  1.70 GB
[feature-1.0.0-SNAPSHOT-runner:8250]        setup:   2,393.27 ms,  1.70 GB
11:42:54,312 INFO  [org.jbo.threads] JBoss Threads version 3.2.0.Final
[feature-1.0.0-SNAPSHOT-runner:8250]     (clinit):     756.77 ms,  5.41 GB
[feature-1.0.0-SNAPSHOT-runner:8250]   (typeflow):  15,306.95 ms,  5.41 GB
[feature-1.0.0-SNAPSHOT-runner:8250]    (objects):  20,260.21 ms,  5.41 GB
[feature-1.0.0-SNAPSHOT-runner:8250]   (features):   1,138.07 ms,  5.41 GB
[feature-1.0.0-SNAPSHOT-runner:8250]     analysis:  39,254.96 ms,  5.41 GB
Error: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Trace: 
        at parsing org.testcontainers.utility.Base58.randomString(Base58.java:19)
Call path from entry point to org.testcontainers.utility.Base58.randomString(int): 
        at org.testcontainers.utility.Base58.randomString(Base58.java:16)
        at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:139)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:90)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:82)
        at ai.tusimple.map.feature.AppLifecycleBean.initLocalS3(AppLifecycleBean.java:69)
        at ai.tusimple.map.feature.AppLifecycleBean.onStart(AppLifecycleBean.java:134)
        at ai.tusimple.map.feature.AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.notify(AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.zig:147)
        at io.quarkus.arc.impl.EventImpl$DeferredEventNotification.run(EventImpl.java:427)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
        at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

com.oracle.svm.core.util.UserError$UserException: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Trace: 
        at parsing org.testcontainers.utility.Base58.randomString(Base58.java:19)
Call path from entry point to org.testcontainers.utility.Base58.randomString(int): 
        at org.testcontainers.utility.Base58.randomString(Base58.java:16)
        at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:139)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:90)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:82)
        at ai.tusimple.map.feature.AppLifecycleBean.initLocalS3(AppLifecycleBean.java:69)
        at ai.tusimple.map.feature.AppLifecycleBean.onStart(AppLifecycleBean.java:134)
        at ai.tusimple.map.feature.AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.notify(AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.zig:147)
        at io.quarkus.arc.impl.EventImpl$DeferredEventNotification.run(EventImpl.java:427)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
        at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

        at com.oracle.svm.core.util.UserError.abort(UserError.java:82)
        at com.oracle.svm.hosted.FallbackFeature.reportAsFallback(FallbackFeature.java:233)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:798)
        at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:582)
        at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:495)
        at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
Detailed message:
Trace: 
        at parsing org.testcontainers.utility.Base58.randomString(Base58.java:19)
Call path from entry point to org.testcontainers.utility.Base58.randomString(int): 
        at org.testcontainers.utility.Base58.randomString(Base58.java:16)
        at org.testcontainers.containers.GenericContainer.<init>(GenericContainer.java:139)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:90)
        at org.testcontainers.containers.localstack.LocalStackContainer.<init>(LocalStackContainer.java:82)
        at ai.tusimple.map.feature.AppLifecycleBean.initLocalS3(AppLifecycleBean.java:69)
        at ai.tusimple.map.feature.AppLifecycleBean.onStart(AppLifecycleBean.java:134)
        at ai.tusimple.map.feature.AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.notify(AppLifecycleBean_Observer_onStart_fd71b5e0b207b7d1ef838b94eaeff75e52b8f463.zig:147)
        at io.quarkus.arc.impl.EventImpl$DeferredEventNotification.run(EventImpl.java:427)
        at java.lang.Thread.run(Thread.java:829)
        at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:553)
        at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:192)
        at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)

        at com.oracle.graal.pointsto.constraints.UnsupportedFeatures.report(UnsupportedFeatures.java:126)
        at com.oracle.svm.hosted.NativeImageGenerator.runPointsToAnalysis(NativeImageGenerator.java:795)
        ... 8 more
Caused by: com.oracle.graal.pointsto.constraints.UnsupportedFeatureException: Detected an instance of Random/SplittableRandom class in the image heap. Instances created during image generation have cached seed values and don't behave as expected.  To see how this object got instantiated use --trace-object-instantiation=java.security.SecureRandom. The object was probably created by a class initializer and is reachable from a static field. You can request class initialization at image runtime by using the option --initialize-at-run-time=<class-name>. Or you can write your own initialization methods and call them explicitly from your main entry point.
        at com.oracle.svm.hosted.image.DisallowedImageHeapObjectFeature.error(DisallowedImageHeapObjectFeature.java:163)
        at com.oracle.svm.core.image.DisallowedImageHeapObjects.check(DisallowedImageHeapObjects.java:65)
        at com.oracle.svm.hosted.image.DisallowedImageHeapObjectFeature.replacer(DisallowedImageHeapObjectFeature.java:139)
        at com.oracle.graal.pointsto.meta.AnalysisUniverse.replaceObject(AnalysisUniverse.java:565)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.replaceObject(AnalysisConstantReflectionProvider.java:213)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.interceptValue(AnalysisConstantReflectionProvider.java:184)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readValue(AnalysisConstantReflectionProvider.java:98)
        at com.oracle.svm.hosted.ameta.AnalysisConstantReflectionProvider.readFieldValue(AnalysisConstantReflectionProvider.java:77)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.util.ConstantFoldUtil$1.readValue(ConstantFoldUtil.java:51)
        at jdk.internal.vm.compiler/org.graalvm.compiler.core.common.spi.JavaConstantFieldProvider.readConstantField(JavaConstantFieldProvider.java:84)
        at com.oracle.svm.hosted.ameta.AnalysisConstantFieldProvider.readConstantField(AnalysisConstantFieldProvider.java:70)
        at jdk.internal.vm.compiler/org.graalvm.compiler.nodes.util.ConstantFoldUtil.tryConstantFold(ConstantFoldUtil.java:47)
        at com.oracle.svm.hosted.phases.ConstantFoldLoadFieldPlugin.tryConstantFold(ConstantFoldLoadFieldPlugin.java:61)
        at com.oracle.svm.hosted.phases.ConstantFoldLoadFieldPlugin.handleLoadStaticField(ConstantFoldLoadFieldPlugin.java:57)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genGetStatic(BytecodeParser.java:4972)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genGetStatic(BytecodeParser.java:4939)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5442)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3451)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3258)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1125)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1019)
        at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
        at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:76)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
        at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
        at com.oracle.graal.pointsto.flow.AnalysisParsedGraph.parseBytecode(AnalysisParsedGraph.java:113)
        at com.oracle.svm.hosted.SVMHost.parseBytecode(SVMHost.java:647)
        at com.oracle.graal.pointsto.meta.AnalysisMethod.ensureGraphParsed(AnalysisMethod.java:592)
        at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:163)
        at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:304)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.createTypeFlow(MethodTypeFlow.java:313)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureTypeFlowCreated(MethodTypeFlow.java:302)
        at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
        at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:420)
        at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
        at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
        ... 5 more
Error: Image build request failed with exit status 1

> Task :feature:quarkusBuild FAILED
:feature:quarkusBuild (Thread[Execution worker for ':',5,main]) completed. Took 46.482 secs.

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':feature:quarkusBuild'.
> io.quarkus.builder.BuildException: Build failure: Build failed due to errors
        [error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Failed to build native image
        at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:209)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:920)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
        at java.base/java.lang.Thread.run(Thread.java:829)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
  Caused by: java.lang.RuntimeException: Image generation failed. Exit code: 1
        at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.imageGenerationFailed(NativeImageBuildStep.java:350)
        at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:181)
        ... 10 more


* Try:
Run with --stacktrace option to get the stack trace. Run with --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 17s
10 actionable tasks: 6 executed, 4 up-to-date
Watched directory hierarchies: [/Users/ziqi.liu/workspace/map-java-service]
```

## To Reproduce

It's from a very large project I haven't found a minimal reproducible example yet

### Configuration
```properties
# Add your application.properties here, if applicable.

```

### Screenshots
(If applicable, add screenshots to help explain your problem.)

## Additional context
(Add any other context about the problem here.)

**Comment by stuartwdouglas** [comment]
You would need to make sure that the testcontainers code is not 'reachable' from native image code, in a way that the native compiler can understand that this is not used and avoid compiling it into the image.

Often the simplest way is a substitution, if you only have one method in your code that is the entry point for your testcontainers stuff you can do something like: https://github.com/quarkusio/quarkus/blob/f85934ea9ecb8d2b9e9ba974e8c31b9a464f7985/extensions/artemis-core/runtime/src/main/java/io/quarkus/artemis/core/runtime/graal/CheckDependenciesSubstitutions.java#L9

Basically define a method that matches your entry point method, then have it not invoke testcontainers (e.g. an empty method), which will make graal stop including it.

Note that this gets a lot more complex if you have references to testcontainers all through your code.

**Comment by liuzqt** [comment]
@stuartwdouglas what does it mean by 'match your entry point method' ? Yes I only have one entry point when the whole app init.

```
@ApplicationScoped
public class AppLifecycleBean {

    private void initLocalS3() {
      // testcontainer staff here...
    }


    void onStart(@Observes StartupEvent ev) {
        LOGGER.info("The application is starting...");
        if (profile.equals("dev") || profile.equals("test")) {
            // only do this in DEV/TEST
            initLocalS3();
        }

    }
}

```


so I probably should make such class:
```
@Substitute
@TargetClass(org.apache.activemq.artemis.core.remoting.impl.netty.CheckDependencies.class)
final class CheckDependenciesSubstitutions {

    @Substitute
    public static final void initLocalS3() {
        return;    }
}
```

I'm not sure what does mean by `match` here, is the class name or method name matters? and how would the native compiler know which method to substitude

**Comment by geoand** [comment]
My question is, why start a container from application code?

**Comment by liuzqt** [comment]
@geoand  What I want to do is to start the localstack container to mock S3 in dev/test mode, but don't really need it in prod and native mode so I don't really need to make it work in native. The substitute approach mentioned by @stuartwdouglas seems to be a good solution for me

**Comment by geoand** [comment]
Then I wonder if this use case is something we want to cover with dev-services... I mean if this is a common case, it would be nice if Quarkus covered it without requiring user code interaction

**Comment by gsmet** [comment]
For tests, you have https://quarkus.io/guides/getting-started-testing#quarkus-test-resource but it won't cover dev mode.

**Comment by stuartwdouglas** [comment]
Part of the issue is we don't really have a 'dev class path', although with continuous testing we do have access to the test class path, so we could abuse that somehow, but I think that has the potential to get very messy.

**Comment by liuzqt** [comment]
@geoand Yeah mocking S3 in local might be as useful as testcontainer datasource which is built in with dev-services.  

The test-resource as mentioned by @gsmet  is already good enough for testing, the most important thing I want is to have a perfectly self-contained testing environment.  

For dev, currently, it's ok for me to have some short scripts to set up things I need instead of mess it up in the quarkus app code. But supporting S3 mocking in dev-services might still be helpful.

**Comment by stuartwdouglas** [comment]
We are looking at adding generic container support to dev services, so you would potentially just be able to configure the image name, ports etc and not need any code. Does this sound useful for your use case?

**Comment by liuzqt** [comment]
@stuartwdouglas  absolutely that would be very useful. We're exploring many testcontainer choices as our local mocking environment in dev/test mode.
